### ISSUE TITLE
watchOS crash, Termination Reason: Code 0xdead10cc

### ISSUE THREAD

**Issue by funcJeff** [Adobe-Marketing-Cloud/mobile-services#352]
I am seeing crashes with this stack trace in thread 0.

```
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x234aa540 mach_msg_trap + 8
1   libsystem_kernel.dylib        	0x234a9a0c mach_msg + 72 (mach_msg.c:103)
2   CoreFoundation                	0x238bfb90 __CFRunLoopServiceMachPort + 212 (CFRunLoop.c:2575)
3   CoreFoundation                	0x238bae8c __CFRunLoopRun + 1760 (CFRunLoop.c:2931)
4   CoreFoundation                	0x238ba4a0 CFRunLoopRunSpecific + 464 (CFRunLoop.c:3192)
5   GraphicsServices              	0x26e2a2ec GSEventRunModal + 100 (GSEvent.c:2246)
6   UIKitCore                     	0x400b3784 UIApplicationMain + 1760 (UIApplication.m:4687)
7   libxpc.dylib                  	0x235a8aa4 _xpc_objc_main.cold.3 + 208
8   libxpc.dylib                  	0x23594a90 _xpc_objc_main + 212 (main.m:108)
9   libxpc.dylib                  	0x235973e4 xpc_main + 152 (init.c:1568)
10  Foundation                    	0x2428a63c -[NSXPCListener resume] + 160 (NSXPCListener.m:276)
11  PlugInKit                     	0x2d02a2f8 -[PKService run] + 396 (PKService.m:165)
12  WatchKit                      	0x35da8f94 WKExtensionMain + 60 (main.m:19)
13  WatchKit                      	0x35da8fb8 main + 12 (main.m:27)
14  libdyld.dylib                 	0x23354a38 start + 4
```

[TN 2151](https://developer.apple.com/library/archive/technotes/tn2151/_index.html) says that reason code  _0xdead10cc_ is because "...application has been terminated by the OS because it held on to a file lock or sqlite database lock during suspension..."

Nothing in the app uses SQLite except ADBMobile, so that must be the culprit. Sure enough, thread 3 has this stack trace.

```
0   libsystem_kernel.dylib        	0x234ce734 guarded_pwrite_np + 12
1   libsqlite3.dylib              	0x23cd0144 unixWrite + 228 (sqlite3.c:38430)
2   libsqlite3.dylib              	0x23d2051c pagerAddPageToRollbackJournal + 412 (sqlite3.c:22798)
3   libsqlite3.dylib              	0x23ccba48 pager_write + 304 (sqlite3.c:63404)
4   libsqlite3.dylib              	0x23ccfe34 sqlite3PagerCommitPhaseOne + 1660 (sqlite3.c:63560)
5   libsqlite3.dylib              	0x23cbb3d0 sqlite3BtreeCommitPhaseOne + 160 (sqlite3.c:73514)
6   libsqlite3.dylib              	0x23c88f08 sqlite3VdbeHalt + 3608 (sqlite3.c:85234)
7   libsqlite3.dylib              	0x23cb3f0c sqlite3VdbeExec + 50532 (sqlite3.c:91464)
8   libsqlite3.dylib              	0x23ca6608 sqlite3_step + 748 (sqlite3.c:88575)
9   QuizboyWatch                 	0x050fbbb0 __38-[ADBAnalyticsWorker queue:timestamp:]_block_invoke + 140
10  libdispatch.dylib             	0x233173a0 _dispatch_call_block_and_release + 24 (init.c:1408)
11  libdispatch.dylib             	0x23318730 _dispatch_client_callout + 16 (object.m:495)
12  libdispatch.dylib             	0x2331e844 _dispatch_lane_serial_drain + 676 (inline_internal.h:2487)
13  libdispatch.dylib             	0x2331f230 _dispatch_lane_invoke + 408 (queue.c:3820)
14  libdispatch.dylib             	0x23328728 _dispatch_workloop_worker_thread + 580 (queue.c:6380)
15  libsystem_pthread.dylib       	0x2354aa0c _pthread_wqthread + 280 (pthread.c:2323)
16  libsystem_pthread.dylib       	0x23550a64 start_wqthread + 8
```

Quizboy is my wrapper to ADBMobile. It assembles some common context variables and so on.

I am not finding any search hits on this problem, and I see no fixes mentioned in the latest release notes (4.18.9). I am using 4.18.2. Therefore, perhaps it is something in Quizboy that is contributing to the problem.

Quizboy wraps `trackAction` and `trackAction` calls and dispatches them to a dispatch queue with a [background quality-of-service](https://developer.apple.com/documentation/dispatch/dispatchqos/qosclass). After all, it is analytics, who cares how fast it gets sent? No offense.

I am quasi-confident that calling `ADBMobile.trackAction` and `ADBMobile.trackState` from a `DispatchQueue` with a `DispatchQoS.QoSClass` value of `background` is causing this problem. My intention is to change the `qos` to `utility` and see if crashes are reduced or eliminated.

I am writing this issue because I would like word from Adobe folks if my reasoning, conclusion, and fix are correct or not, and to document this problem for other devs that throw ADBMobile calls to low priority worker threads.

**Comment by erf23** [comment]
Hi @funcJeff -
First off, thank you for bringing this issue to my (and the rest of the team's) attention and providing the useful details and tips. I also personally want to apologize for the delay in replying to you.

From my standpoint and with the information you've gathered, your reasoning, conclusion, and workaround make sense to me. Since dispatch queues and qos are involved, I think it is definitely possible we have some lurking issues there.

To kick off the process, I'm going to file an internal ticket. The ticket will help us track and prioritize this issue appropriately, and makes sure we follow through on resolving this.

To help me with that, I have a few questions (if you have time):
- I'd like to understand if this issue is still affecting you. How negatively is this impacting you? Are you blocked?
- I'm also interested in the results of your workaround.
  - Did it work?
  - Did something else work.
- If you've got any other information, I'd love to hear it.

Concurrently, I will get our team started on doing our side's homework with the information you've already provided. I'm not sure if this issue will end up on my plate, but I plan to remain involved in the process. Again, thank you and if you have questions, just reach out.

Glenn Eguchi
Adobe Mobile SDK Engineer

**Comment by funcJeff** [comment]
@erf23 

> How negatively is this impacting you? 

This crash only happens as the Watch app goes into the background. Right now, our app does **not** have dynamic state it needs to preserve when entering background, for example, like Workout or Timer does. Thus, the user does not notice this crash. So our impact in minimal.

> Did it [the fix] work?

No. The fix I proposed above has been released for 5 days in production now. It did **not** fix the problem. This crash is our #2 crash right now.

```
Hardware Model:      Watch4,3
Process:             PinkShoeWatch Extension [1333]
Path:                /private/var/containers/Bundle/Application/scrubbed/PinkShoeWatch.app/PlugIns/PinkShoeWatch Extension.appex/PinkShoeWatch Extension
Identifier:          com.pinkshoe.iphone.watchkitapp.watchkitextension
Version:             347 (7.24.2)
AppVariant:          1:Watch4,3:6
Code Type:           ARM (Native)
Role:                unknown
Parent Process:      launchd [1]
Coalition:           com.pinkshoe.iphone.watchkitapp.watchkitextension [509]


Date/Time:           2019-11-21 11:12:59.8580 -0800
Launch Time:         2019-11-21 07:57:24.0000 -0800
OS Version:          Watch OS 6.1 (17S84)
Release Type:        User
Baseband Version:    2.01.04
Report Version:      104

Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: Namespace RUNNINGBOARD, Code 0xdead10cc
Triggered by Thread:  0

Thread 0 name:
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x1f1be560 mach_msg_trap + 8
1   libsystem_kernel.dylib        	0x1f1bda2c mach_msg + 72 (mach_msg.c:103)
2   CoreFoundation                	0x1f5d6d3c __CFRunLoopServiceMachPort + 212 (CFRunLoop.c:2575)
3   CoreFoundation                	0x1f5d2038 __CFRunLoopRun + 1760 (CFRunLoop.c:2931)
4   CoreFoundation                	0x1f5d164c CFRunLoopRunSpecific + 464 (CFRunLoop.c:3192)
5   GraphicsServices              	0x22b582ec GSEventRunModal + 100 (GSEvent.c:2246)
6   UIKitCore                     	0x3c319474 UIApplicationMain + 1760 (UIApplication.m:4758)
7   libxpc.dylib                  	0x1f2bc9f0 _xpc_objc_main.cold.3 + 208
8   libxpc.dylib                  	0x1f2a8948 _xpc_objc_main + 212 (main.m:126)
9   libxpc.dylib                  	0x1f2ab29c xpc_main + 152 (init.c:1568)
10  Foundation                    	0x1ffa572c -[NSXPCListener resume] + 160 (NSXPCListener.m:276)
11  PlugInKit                     	0x28deaef0 -[PKService run] + 396 (PKService.m:165)
12  WatchKit                      	0x31c131f4 WKExtensionMain + 60 (main.m:19)
13  WatchKit                      	0x31ce18e0 main + 12 (main.m:27)
14  libdyld.dylib                 	0x1f064a78 start + 4

Thread 1 name:
Thread 1:
0   libsystem_kernel.dylib        	0x1f1be59c semaphore_wait_trap + 8
1   libdispatch.dylib             	0x1f028ab4 _dispatch_sema4_wait + 24 (lock.c:139)
2   libdispatch.dylib             	0x1f0290f8 _dispatch_semaphore_wait_slow + 128 (semaphore.c:132)
3   Foundation                    	0x1ffe4ea8 -[NSFileCoordinator(NSPrivate) _blockOnAccessClaim:withAccessArbiter:] + 364 (NSFileCoordinator.m:843)
4   Foundation                    	0x1ffcb334 -[NSFileCoordinator(NSPrivate) _coordinateReadingItemAtURL:options:error:byAccessor:] + 624 (NSFileCoordinator.m:1052)
5   Foundation                    	0x1ffcb09c -[NSFileCoordinator coordinateReadingItemAtURL:options:error:byAccessor:] + 92 (NSFileCoordinator.m:494)
6   WatchConnectivity             	0x3300f0e4 -[WCContentIndex loadContentIfNecessary] + 680 (WCContentIndex.m:66)
7   Foundation                    	0x2013f964 __74-[NSFilePresenterXPCMessenger _makePresenter:observeChangeWithSubitemURL:]_block_invoke_2 + 48 (NSFileAccessArbiter.m:3456)
8   Foundation                    	0x2015fd58 __95-[NSOperationQueue(NSFileCoordinationSupport) _fc_addUncancellableOperationForReactorID:block... + 132 (NSFileCoordinator.m:2519)
9   Foundation                    	0x20073d30 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16 (NSOperation.m:1541)
10  Foundation                    	0x1ff806b4 -[NSBlockOperation main] + 100 (NSOperation.m:1560)
11  Foundation                    	0x20075f7c __NSOPERATION_IS_INVOKING_MAIN__ + 20 (NSOperation.m:2184)
12  Foundation                    	0x1ff8034c -[NSOperation start] + 724 (NSOperation.m:2201)
13  Foundation                    	0x2007694c __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20 (NSOperation.m:2215)
14  Foundation                    	0x2007642c __NSOQSchedule_f + 180 (NSOperation.m:2226)
15  libdispatch.dylib             	0x1f027218 _dispatch_call_block_and_release + 24 (init.c:1408)
16  libdispatch.dylib             	0x1f0285a8 _dispatch_client_callout + 16 (object.m:495)
17  libdispatch.dylib             	0x1f02acc8 _dispatch_continuation_pop + 496 (inline_internal.h:2484)
18  libdispatch.dylib             	0x1f02a3e0 _dispatch_async_redirect_invoke + 596 (queue.c:803)
19  libdispatch.dylib             	0x1f03729c _dispatch_root_queue_drain + 356 (inline_internal.h:2525)
20  libdispatch.dylib             	0x1f037a38 _dispatch_worker_thread2 + 120 (queue.c:6628)
21  libsystem_pthread.dylib       	0x1f25e9bc _pthread_wqthread + 216 (pthread.c:2336)
22  libsystem_pthread.dylib       	0x1f264a64 start_wqthread + 8

Thread 2:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 3:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 4 name:
Thread 4:
0   libsystem_kernel.dylib        	0x1f1be560 mach_msg_trap + 8
1   libsystem_kernel.dylib        	0x1f1bda2c mach_msg + 72 (mach_msg.c:103)
2   CoreFoundation                	0x1f5d6d3c __CFRunLoopServiceMachPort + 212 (CFRunLoop.c:2575)
3   CoreFoundation                	0x1f5d2038 __CFRunLoopRun + 1760 (CFRunLoop.c:2931)
4   CoreFoundation                	0x1f5d164c CFRunLoopRunSpecific + 464 (CFRunLoop.c:3192)
5   Foundation                    	0x1ff69e04 -[NSRunLoop(NSRunLoop) runMode:beforeDate:] + 228 (NSRunLoop.m:374)
6   Foundation                    	0x1ff69ce4 -[NSRunLoop(NSRunLoop) runUntilDate:] + 88 (NSRunLoop.m:421)
7   UIKitCore                     	0x3c3a879c -[UIEventFetcher threadMain] + 152 (UIEventFetcher.m:637)
8   Foundation                    	0x2009782c __NSThread__start__ + 844 (NSThread.m:724)
9   libsystem_pthread.dylib       	0x1f25cca0 _pthread_start + 124 (pthread.c:895)
10  libsystem_pthread.dylib       	0x1f264a70 thread_start + 8

Thread 5:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 6:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 7 name:
Thread 7:
0   libsystem_kernel.dylib        	0x1f1be5b4 semaphore_timedwait_trap + 8
1   libdispatch.dylib             	0x1f028b30 _dispatch_sema4_timedwait + 64 (lock.c:154)
2   libdispatch.dylib             	0x1f0290c0 _dispatch_semaphore_wait_slow + 72 (semaphore.c:116)
3   QuizboyWatch                 	0x04fed674 -[ADBNetworkObject sendSync] + 296
4   QuizboyWatch                 	0x04ff06f8 -[ADBAnalyticsWorker workerThread] + 1296
5   QuizboyWatch                 	0x0500386c __30-[ADBHitQueueBase bringOnline]_block_invoke + 72
6   libdispatch.dylib             	0x1f027218 _dispatch_call_block_and_release + 24 (init.c:1408)
7   libdispatch.dylib             	0x1f0285a8 _dispatch_client_callout + 16 (object.m:495)
8   libdispatch.dylib             	0x1f02e7ac _dispatch_lane_serial_drain + 916 (inline_internal.h:2470)
9   libdispatch.dylib             	0x1f02f0a8 _dispatch_lane_invoke + 408 (queue.c:3863)
10  libdispatch.dylib             	0x1f0385d0 _dispatch_workloop_worker_thread + 580 (queue.c:6445)
11  libsystem_pthread.dylib       	0x1f25e9fc _pthread_wqthread + 280 (pthread.c:2323)
12  libsystem_pthread.dylib       	0x1f264a64 start_wqthread + 8

Thread 8:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 9 name:
Thread 9:
0   libsystem_kernel.dylib        	0x1f1e2248 fsync + 8
1   libsqlite3.dylib              	0x1f9e7eec unixSync + 208 (sqlite3.c:38708)
2   libsqlite3.dylib              	0x1f9f1a20 syncJournal + 516 (sqlite3.c:22822)
3   libsqlite3.dylib              	0x1f9e7868 sqlite3PagerCommitPhaseOne + 1288 (sqlite3.c:63935)
4   libsqlite3.dylib              	0x1f9d2f78 sqlite3BtreeCommitPhaseOne + 160 (sqlite3.c:73522)
5   libsqlite3.dylib              	0x1f9a0ab0 sqlite3VdbeHalt + 3608 (sqlite3.c:85242)
6   libsqlite3.dylib              	0x1f9cbab4 sqlite3VdbeExec + 50532 (sqlite3.c:91472)
7   libsqlite3.dylib              	0x1f9be1b0 sqlite3_step + 748 (sqlite3.c:88583)
8   QuizboyWatch                 	0x04fefba0 __38-[ADBAnalyticsWorker queue:timestamp:]_block_invoke + 140
9   libdispatch.dylib             	0x1f027218 _dispatch_call_block_and_release + 24 (init.c:1408)
10  libdispatch.dylib             	0x1f0285a8 _dispatch_client_callout + 16 (object.m:495)
11  libdispatch.dylib             	0x1f02e6bc _dispatch_lane_serial_drain + 676 (inline_internal.h:2484)
12  libdispatch.dylib             	0x1f02f0a8 _dispatch_lane_invoke + 408 (queue.c:3863)
13  libdispatch.dylib             	0x1f0385d0 _dispatch_workloop_worker_thread + 580 (queue.c:6445)
14  libsystem_pthread.dylib       	0x1f25e9fc _pthread_wqthread + 280 (pthread.c:2323)
15  libsystem_pthread.dylib       	0x1f264a64 start_wqthread + 8

Thread 10 name:
Thread 10:
0   libsystem_kernel.dylib        	0x1f1e30fc read + 12
1   Foundation                    	0x2007a568 _NSReadFromFileDescriptorWithProgress + 212 (NSPlatform.m:531)
2   Foundation                    	0x2007ad6c _NSReadBytesFromFileWithExtendedAttributes + 868 (NSPlatform.m:782)
3   Foundation                    	0x1ff6f68c -[NSData(NSData) initWithContentsOfFile:options:maxLength:error:] + 124 (NSData.m:850)
4   Foundation                    	0x1ffabbb8 +[NSData(NSData) dataWithContentsOfFile:options:error:] + 68 (NSData.m:765)
5   WatchConnectivity             	0x330197a0 -[WCFileStorage enumerateIncomingUserInfosWithBlock:] + 904 (WCFileStorage.m:431)
6   WatchConnectivity             	0x33009470 __49-[WCSession handleIncomingUserInfoWithPairingID:]_block_invoke_2 + 316 (WCSession.m:1397)
7   Foundation                    	0x20073d30 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16 (NSOperation.m:1541)
8   Foundation                    	0x1ff806b4 -[NSBlockOperation main] + 100 (NSOperation.m:1560)
9   Foundation                    	0x20075f7c __NSOPERATION_IS_INVOKING_MAIN__ + 20 (NSOperation.m:2184)
10  Foundation                    	0x1ff8034c -[NSOperation start] + 724 (NSOperation.m:2201)
11  Foundation                    	0x2007694c __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20 (NSOperation.m:2215)
12  Foundation                    	0x2007642c __NSOQSchedule_f + 180 (NSOperation.m:2226)
13  libdispatch.dylib             	0x1f027218 _dispatch_call_block_and_release + 24 (init.c:1408)
14  libdispatch.dylib             	0x1f0285a8 _dispatch_client_callout + 16 (object.m:495)
15  libdispatch.dylib             	0x1f02acc8 _dispatch_continuation_pop + 496 (inline_internal.h:2484)
16  libdispatch.dylib             	0x1f02a3e0 _dispatch_async_redirect_invoke + 596 (queue.c:803)
17  libdispatch.dylib             	0x1f03729c _dispatch_root_queue_drain + 356 (inline_internal.h:2525)
18  libdispatch.dylib             	0x1f037a38 _dispatch_worker_thread2 + 120 (queue.c:6628)
19  libsystem_pthread.dylib       	0x1f25e9bc _pthread_wqthread + 216 (pthread.c:2336)
20  libsystem_pthread.dylib       	0x1f264a64 start_wqthread + 8

Thread 11:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 12:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 13 name:
Thread 13:
0   libsystem_kernel.dylib        	0x1f1e2248 fsync + 8
1   Foundation                    	0x2007b9a4 _NSWriteDataToFileWithExtendedAttributes + 616 (NSPlatform.m:1077)
2   WatchConnectivity             	0x33017188 -[WCFileStorage persistOutgoingUserInfoTransfer:] + 472 (WCFileStorage.m:153)
3   WatchConnectivity             	0x32ffe088 -[WCSession onqueue_addOutstandingUserInfoTransfer:] + 208 (WCSession.m:427)
4   WatchConnectivity             	0x33002e34 -[WCSession onqueue_transferUserInfo:withUserInfo:] + 284 (WCSession.m:889)
5   Foundation                    	0x20073d30 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__ + 16 (NSOperation.m:1541)
6   Foundation                    	0x1ff806b4 -[NSBlockOperation main] + 100 (NSOperation.m:1560)
7   Foundation                    	0x20075f7c __NSOPERATION_IS_INVOKING_MAIN__ + 20 (NSOperation.m:2184)
8   Foundation                    	0x1ff8034c -[NSOperation start] + 724 (NSOperation.m:2201)
9   Foundation                    	0x2007694c __NSOPERATIONQUEUE_IS_STARTING_AN_OPERATION__ + 20 (NSOperation.m:2215)
10  Foundation                    	0x2007642c __NSOQSchedule_f + 180 (NSOperation.m:2226)
11  libdispatch.dylib             	0x1f034a70 _dispatch_block_async_invoke2 + 104 (queue.c:525)
12  libdispatch.dylib             	0x1f0285a8 _dispatch_client_callout + 16 (object.m:495)
13  libdispatch.dylib             	0x1f02acc8 _dispatch_continuation_pop + 496 (inline_internal.h:2484)
14  libdispatch.dylib             	0x1f02a3e0 _dispatch_async_redirect_invoke + 596 (queue.c:803)
15  libdispatch.dylib             	0x1f03729c _dispatch_root_queue_drain + 356 (inline_internal.h:2525)
16  libdispatch.dylib             	0x1f037a38 _dispatch_worker_thread2 + 120 (queue.c:6628)
17  libsystem_pthread.dylib       	0x1f25e9bc _pthread_wqthread + 216 (pthread.c:2336)
18  libsystem_pthread.dylib       	0x1f264a64 start_wqthread + 8

Thread 14:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 15:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 16:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 17:
0   libsystem_pthread.dylib       	0x1f264a5c start_wqthread + 0

Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000010004005   x1: 0x0000000007000806   x2: 0x0000000000000000   x3: 0x0000000000000c00
    x4: 0x0000000000001b03   x5: 0x00000000ffffffff   x6: 0x0000000000000000   x7: 0x0000000000000000
    x8: 0x00000000fffffbbf   x9: 0x0000000007000906  x10: 0x00000000d73e000a  x11: 0x00000000d73e000a
   x12: 0x0000000000000000  x13: 0xffffffffffffffff  x14: 0x0000000000000000  x15: 0x04d3b6f348adf870
   x16: 0xffffffffffffffe1  x17: 0x0000000000000000  x18: 0x0000000000000000  x19: 0x0000000000000000
   x20: 0x00000000ffffffff  x21: 0x0000000000001b03  x22: 0x0000000000000c00  x23: 0x0000000004fd2be4
   x24: 0x0000000007000806  x25: 0x0000000000000000  x26: 0x0000000007000806  x27: 0x0000000000000c00
   x28: 0x0000000000000001   fp: 0x0000000004fd2b00   lr: 0x000000001f1bda2c
    sp: 0x0000000004fd2ab0   pc: 0x000000001f1be560 cpsr: 0x60000000
   esr: 0x56000080  Address size fault
```

**Comment by himanshuphalak** [comment]
Hi @funcJeff, 

Firstly, sorry for delay in getting back to you. I wanted to provide you with an update on this ticket.

So far, we have been trying to execute various scenarios and reproduce the issue internally but we haven't had any luck. We also tried running track methods from `DispatchQueue` with a `QOS_CLASS_BACKGROUND` and trying to suspend the app in background. Since we haven't been able to reproduce the exact scenario, I have some follow-up questions if you could please help with:

1. By any chance are you able to reproduce the issue? If so, can you provide some reproducible steps?
2. Can you provide additional technical details on how Quizboy wrapper wraps the ADBMobile APIs? 
3. Looking at the error code, it might also be due to a file lock being open during suspension. Just curious to know if you have already checked off that condition and that the application or any library used by app is not handling any files? 

It would be great if you could share further information around above questions. 

Regards,
Himanshu
Adobe Mobile SDK Engineer

**Comment by funcJeff** [comment]
@himanshuphalak wrote:
> Can you provide additional technical details on how Quizboy wrapper wraps the ADBMobile APIs

Here is a sanitized and simplified version with all pertinent details.

```
import Foundation

public protocol QTracker {
    var stateName:String { get }
    func track(_ event:QTrackEvent, context:[String:String]?)
}

public extension QTracker {
    func track(_ event:QTrackEvent, context:[String:String]? = nil) {
        Quizboy.track(event, stateName: stateName, context: context ?? [:])
    }
}

public enum QTrackEvent {
    case state
    case action(String)
}

public protocol QContext {
    var name:String { get }
    var value:String { get }
}

internal protocol ADBMWrappable {
    func trackAction(_ action: String, data: [String : String])
    func trackState(_ state: String, data: [String : String])
}

internal struct ADBMWrapper: ADBMWrappable {
    func trackAction(_ action: String, data: [String : String]) {
        ADBMobile.trackAction(action, data: data)
    }
    func trackState(_ state: String, data: [String : String]) {
        ADBMobile.trackState(state, data: data)
    }
}

public struct Quizboy {
    public static var appName:String = "APPNAME"
    internal static var wrapper:ADBMWrappable = ADBMWrapper()
    public static var globalContext:[QContext] = []
    private static var serialQueue = DispatchQueue(label: "QuizboyQueue", qos: .utility, attributes: [], autoreleaseFrequency:.inherit, target:DispatchQueue.global(qos: .utility))
    private static var logQueue = DispatchQueue(label: "QuizboyLogQueue", qos: .utility, attributes: [], autoreleaseFrequency:.inherit, target:DispatchQueue.global(qos: .utility))
    public static var logDebug:Bool = false
    internal static var qLogger:QLogable = QLogger()
    private static let defaultContext = [ "fruit":"apple"]
    
    private static func addActionContext(_ context: inout [String:String], _ actionName:String, _ stateName:String, _ appName:String) {
        // String & dictionary munging.
    }
    
    private static func addStateContext(_ context: inout [String:String], _ stateName:String, _ appName:String) {
        // String & dictionary munging.
    }
    
    internal static func track(_ event:QTrackEvent, stateName:String, context:[String:String] = [:]) {
        serialQueue.async {
            var fullContext:[String:String] = defaultContext
            
            switch event {
            case .action(let actionName):
                addActionContext(&fullContext, actionName, stateName, appName)
            case .state:
                addStateContext(&fullContext, stateName, appName)
            }
            
            globalContext.forEach{ fullContext[$0.name] = $0.value }
            context.forEach{ (key, value) in fullContext[key] = value }
            
            if logDebug {
                logQueue.async {
                    self.qLogger.log(event:event, for:stateName, withContext:fullContext)
                }
            }
            
            switch event {
            case .action(let name):
                wrapper.trackAction(name, data: fullContext)
            case .state:
                wrapper.trackState(stateName, data: fullContext)
            }
        }
    }
}

internal protocol QLogable {
    func log(event:QTrackEvent, for stateName:String, withContext context:[String:String])
}

internal struct QLogger: QLogable {
    func log(event:QTrackEvent, for stateName:String, withContext context:[String:String]) {
        // Print output.
    }
}
```

And here is how it is used.

```
import WatchKit
import Foundation

class InterfaceController: WKInterfaceController, QTracker {
    var stateName = "CoolScene"
    
    override func didAppear() {
        track(.state)
    }
    
    @IBAction func didTapGo() {
        track(.action("Go"))
    }
}
```

**Comment by funcJeff** [comment]
@himanshuphalak wrote:
> ...it might also be due to a file lock being open during suspension ... if you have already checked off that condition

The file write command our code uses is [`Data.write(to:options:)`](https://developer.apple.com/documentation/foundation/data/1779858-write).

The file read command our code uses is [`Data.init(contentsOf:options:)`](https://developer.apple.com/documentation/foundation/data/3126626-init).

To fetch the URL for those files, our code uses [`FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)`](https://developer.apple.com/documentation/foundation/filemanager/1407726-urls).

To delete items, our code uses [`FileManager.default.removeItem(at:)`](https://developer.apple.com/documentation/foundation/filemanager/1413590-removeitem)

None of the docs for those funcs mention anything about file locking. I am not certain, but I doubt such simple read/write operations use any kind of file locking.

> or any library used by app is not handling any files?

ADBMobile is the only library our Watch extension is using.

Further, every stack trace has threads like this in it:

```
Thread 8 name:
Thread 8:
0   libsystem_platform.dylib      	0x1a760b7c __bzero + 28
1   libsqlite3.dylib              	0x1aeae218 purgeableCacheRetainChunk + 128 (sqlite3.c:55745)
2   libsqlite3.dylib              	0x1af87810 purgeableCachePagePin + 172 (sqlite3.c:56076)
3   libsqlite3.dylib              	0x1aeadd28 purgeableCacheFetch + 104 (sqlite3.c:56444)
4   libsqlite3.dylib              	0x1af3eecc getPageNormal + 72 (sqlite3.c:53379)
5   libsqlite3.dylib              	0x1aea99f8 sqlite3BtreeBeginTrans + 536 (sqlite3.c:63063)
6   libsqlite3.dylib              	0x1aec8324 sqlite3VdbeExec + 3540 (sqlite3.c:93696)
7   libsqlite3.dylib              	0x1aec61b0 sqlite3_step + 748 (sqlite3.c:88583)
8   Quizboy                 	        0x00b4bba0 __38-[ADBAnalyticsWorker queue:timestamp:]_block_invoke + 140
9   libdispatch.dylib             	0x1a52f218 _dispatch_call_block_and_release + 24 (init.c:1408)
10  libdispatch.dylib             	0x1a5305a8 _dispatch_client_callout + 16 (object.m:495)
11  libdispatch.dylib             	0x1a5366bc _dispatch_lane_serial_drain + 676 (inline_internal.h:2484)
12  libdispatch.dylib             	0x1a5370a8 _dispatch_lane_invoke + 408 (queue.c:3863)
13  libdispatch.dylib             	0x1a5405d0 _dispatch_workloop_worker_thread + 580 (queue.c:6445)
14  libsystem_pthread.dylib       	0x1a7669fc _pthread_wqthread + 280 (pthread.c:2323)
15  libsystem_pthread.dylib       	0x1a76ca64 start_wqthread + 8

Thread 9 name:
Thread 9:
0   libsystem_kernel.dylib        	0x1a6c65b4 semaphore_timedwait_trap + 8
1   libdispatch.dylib             	0x1a530b30 _dispatch_sema4_timedwait + 64 (lock.c:154)
2   libdispatch.dylib             	0x1a5310c0 _dispatch_semaphore_wait_slow + 72 (semaphore.c:116)
3   Quizboy                 	        0x00b49674 -[ADBNetworkObject sendSync] + 296
4   Quizboy                       	0x00b4c6f8 -[ADBAnalyticsWorker workerThread] + 1296
5   Quizboy                 	        0x00b5f86c __30-[ADBHitQueueBase bringOnline]_block_invoke + 72
6   libdispatch.dylib             	0x1a52f218 _dispatch_call_block_and_release + 24 (init.c:1408)
7   libdispatch.dylib             	0x1a5305a8 _dispatch_client_callout + 16 (object.m:495)
8   libdispatch.dylib             	0x1a5367ac _dispatch_lane_serial_drain + 916 (inline_internal.h:2470)
9   libdispatch.dylib             	0x1a5370a8 _dispatch_lane_invoke + 408 (queue.c:3863)
10  libdispatch.dylib             	0x1a5405d0 _dispatch_workloop_worker_thread + 580 (queue.c:6445)
11  libsystem_pthread.dylib       	0x1a7669fc _pthread_wqthread + 280 (pthread.c:2323)
12  libsystem_pthread.dylib       	0x1a76ca64 start_wqthread + 8
```

Those 2 threads make it clear that one is doing something in SQLite and the other is waiting on some condition in another thread.

We have 100+ different stack traces with similar conditions.

**Comment by funcJeff** [comment]
@himanshuphalak wrote:

> By any chance are you able to reproduce the issue?

I have not attempted to reproduce the problem. I am not certain what I would do if I did reproduce it.

> can you provide some reproducible steps?

Based on the Termination Reason given in the crash report, and what [TN 2151](https://developer.apple.com/library/archive/technotes/tn2151/_index.html) tells us that means, try putting the app in the background over and over again. I would go about it like this:

1. From Home screen, tap your app icon
1. Your app launches and is running in the foreground
1. Wait a few seconds
1. Deep press Watch Display
1. Context Menu appears
1. Press button on Context Menu
1. Your app's main interface reappears
1. Press the Side Button
1. Turn Digital Crown in the Dock to show a different app
1. In the Doc, tap the different app's card
1. Your app is suspended, the different app is running in the foreground
1. Wait a few seconds
1. Press the Side Button
1. Turn Digital Crown in the Dock to show your app
1. In the Doc, tap your app's card
1. The different app is suspended, your app is running in the foreground
1. Go to step 3; repeat 100 times

Obviously, your app should call `ADBMobile.trackState` every time the `WKInterfaceController` appears as well as `ADBMobile.trackAction` for context menu items. For example:

```
class CherryInterfaceController: WKInterfaceController {
    
    override func awake(withContext context: Any?) {
        addMenuItem(with: WKMenuItemIcon.resume, title: "Reload", action: #selector(reload))
    }
    
    override func didAppear() {
        ADBMobile.trackState("Cherry", data:stateData)
    }
    
    @objc func reload() {
        ADBMobile.trackAction("Reload", data:actionData)
    }
}
```

Since we are talking about reproducing the crash, here are some notes on frequency.

We have had 186 occurrences in the last 7 days. This crash is the top crash for our app in Xcode Organizer. This means it is more frequent than the top iOS crash. We have 1000 times more iOS users than watchOS users. Thus, the per capita crash frequency is extreme.

**Comment by funcJeff** [comment]
@himanshuphalak

I have been reading the docs on [watchOS app life cycles](https://developer.apple.com/documentation/watchkit/working_with_the_watchos_app_life_cycle), and I think you should add a method to `ADBMobile`. Maybe name it `databaseStop` or similar. This new method will immediately cancel all SQLLite operations. When my app's `WKExtensionDelegate` is called via `applicationWillResignActive` then my app will call `ADBMobile.databaseStop`. Unlike [UIApplication](https://developer.apple.com/documentation/uikit/uiapplication/1622973-willresignactivenotification) there are no global notifications of app state change on watchOS.
